crates/kitsune2_showcase/src/app/mod.rs

```diff
@@ -0,0 +1,391 @@
+        )
+        .await?
+        .into_iter()
+        .filter(|p| p.url.is_some())
+        .collect::<Vec<_>>();
+        self.printer_tx
+            .send(format!("Publishing op to {} peers", peers.len()))
+            .await
+            .expect("Failed to print message");
+
+        for peer in peers {
+            let printer_tx = self.printer_tx.clone();
+            let space = self.space.clone();
+            let op_id = op_id.clone();
+            if let Some(url) = peer.url.clone() {
+                tokio::task::spawn(async move {
+                    match space.publish().publish_ops(vec![op_id], url).await {
+                        Ok(_) => printer_tx
+                            .send(format!("Published to {}", &peer.agent))
+                            .await
+                            .expect("Failed to print message"),
+                        Err(err) => printer_tx
+                            .send(format!(
+                                "Failed to publish to {}: {err:?}",
+                                &peer.agent
+                            ))
+                            .await
+                            .expect("Failed to print message"),
+                    }
+                });
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn list(&self) -> K2Result<()> {
+        let op_ids = self
+            .space
+            .op_store()
+            .retrieve_op_hashes_in_time_slice(
+                DhtArc::FULL,
+                Timestamp::from_micros(0),
+                Timestamp::now(),
+            )
+            .await?
+            .0;
+
+        let stored_ops =
+            self.space.op_store().retrieve_ops(op_ids.to_vec()).await?;
+
+        if !stored_ops.is_empty() {
+            self.printer_tx
+                .send("NAME\t\t\t\t\t\tCREATED AT\t\t\t\tID".to_string())
+                .await
+                .expect("Failed to print message");
+
+            for op in stored_ops {
+                let mem_op = MemoryOp::from(op.op_data);
+                let created_at = DateTime::<Local>::from(SystemTime::from(
+                    mem_op.created_at,
+                ));
+                let file_name =
+                    serde_json::from_slice::<FileData>(&mem_op.op_data)
+                        .map(|f| f.name)
+                        .unwrap_or("<CORRUPTED FILE DATA>".to_string());
+                self.printer_tx
+                    .send(format!(
+                        "{:40}\t{:?}\t{}",
+                        file_name, created_at, op.op_id
+                    ))
+                    .await
+                    .expect("Failed to print message");
+            }
+        } else {
+            self.printer_tx
+                .send("No ops found".to_string())
+                .await
+                .expect("Failed to print message");
+        }
+
+        Ok(())
+    }
+
+    pub async fn fetch(&self, file_name: &str) -> K2Result<()> {
+        let op_ids = self
```

Comment:
```suggestion
    pub async fn fetch(&self, file_name: &str) -> K2Result<()> {
        // This demo application is pulling the full list of local
        // ops in order to find the file name. In a real application
        // it would be more efficient to store the file at its CAS
        // address, then have a link stored at the hash of the
        // filename, so you could look up a single entry to find
        // the CAS address of the file data.
        let op_ids = self
```